core/src/lib.rs

```diff
@@ -627,6 +631,51 @@ impl PGMQueue {
         Ok(messages)
     }
 
+    /// Similar to [`read_batch`], but allows waiting until a message is available
+    ///
+    /// You can specify a maximum duration for polling (defaults to 5 seconds),
+    /// and an interval between calls (defaults to 250ms). A lower interval
+    /// implies higher maximum latency, but less load on the database.
+    ///
+    /// Refer to the [`read_batch`] function for more details.
+    ///
+    pub async fn read_batch_with_poll<T: for<'de> Deserialize<'de>>(
+        &self,
+        queue_name: &str,
+        vt: Option<i32>,
+        max_batch_size: i32,
+        poll_timeout: Option<Duration>,
+        poll_interval: Option<Duration>,
+    ) -> Result<Option<Vec<Message<T>>>, errors::PgmqError> {
+        let vt_ = vt.unwrap_or(VT_DEFAULT);
+        let poll_timeout_ = poll_timeout.unwrap_or(POLL_TIMEOUT_DEFAULT);
+        let poll_interval_ = poll_interval.unwrap_or(POLL_INTERVAL_DEFAULT);
+        let start_time = std::time::Instant::now();
+        loop {
+            let query = &query::read(queue_name, vt_, max_batch_size)?;
+            let messages = fetch_messages::<T>(query, &self.connection).await?;
+            // Why can fetch_messages return both `None` and `Option(())`
+            match messages {
+                None => {
+                    if start_time.elapsed() < poll_timeout_ {
+                        tokio::time::sleep(poll_interval_).await;
+                        continue;
+                    } else {
+                        break Ok(None);
+                    }
+                }
+                Some(m) => {
+                    if m.is_empty() && start_time.elapsed() < poll_timeout_ {
+                        tokio::time::sleep(poll_interval_).await;
+                        continue;
+                    } else {
+                        break Ok(Some(m));
+                    }
+                }
+            }
```

Comment:
This looks awkward, accepting suggestions on how to improve it

Reply:
It's probably awkward because fetch_messages() is a little messy.

fetch_messages() is supposed to return `Some(Vec<Messages>)` if there are messages to return, otherwise it should be `None`. It shouldn't every return an empty Vec, unless there's a bug. Let me look at fetch_messages() and see if that signature can be made more explicit.

Reply:
As discussed, it was needed due to a bug in fetch_messages. Fixed bug in https://github.com/tembo-io/pgmq/pull/64/commits/a3dc314dfed8a6142f361958ec9410d55af971f8